Penteract/Assets/Scripts/FloorIsLava.cpp

```diff
@@ -0,0 +1,208 @@
+#include "FloorIsLava.h"
+
+#include "GameObject.h"
+#include "GameplaySystems.h"
+
+#include <math.h>
+
+EXPOSE_MEMBERS(FloorIsLava) {
+    MEMBER(MemberType::GAME_OBJECT_UID, corridorUID),
+    MEMBER(MemberType::GAME_OBJECT_UID, arenaUID),
+	MEMBER(MemberType::FLOAT, timeTilesActive),
+	MEMBER(MemberType::FLOAT, timeWarning)
+
+};
+
+GENERATE_BODY_IMPL(FloorIsLava);
+
+void FloorIsLava::Start() {
+	
+	corridor = GameplaySystems::GetGameObject(corridorUID);
+	if (corridor) {
+		int i = 0;
+		for (GameObject* children : corridor->GetChildren()) {
+			corridorTiles[i] = children;
+			++i;
+		}
+	}
+	arena = GameplaySystems::GetGameObject(arenaUID);
+	if (arena) {
+		int i = 0;
+		for (GameObject* children : arena->GetChildren()) {
+			arenaTiles[i] = children;
+			++i;
+		}
+	}
+
+	timeRemainingWarning = timeWarning;
+	currentPattern = previousPattern;
+}
+
+void FloorIsLava::Update() {
+	
+	//select a random corridor and arena pattern
+	if (patternFinished) {
+		while (currentPattern == previousPattern) {
+			currentPattern = 1 + (rand() % 3);
+		}
+		switch (currentPattern)
+		{
+		case 1:
+			currentCorridorPattern = corridorPattern1;
+			currentArenaPattern = arenaPattern1;
+			break;
+		case 2:
+			currentCorridorPattern = corridorPattern2;
+			currentArenaPattern = arenaPattern2;
+			break;
+		case 3:
+			currentCorridorPattern = corridorPattern3;
+			currentArenaPattern = arenaPattern3;
+			break;
+		default:
+			break;
+		}
+	}
+	
+	if (warningActive) {
+		if (firstTimeWarning) {
+			UpdateWarningTiles(true);
+			firstTimeWarning = false;
+		}
+
+		if (timeRemainingWarning > 0.f) {
+			timeRemainingWarning -= Time::GetDeltaTime();
+		}
+		else {
+			UpdateWarningTiles(false);
+			warningActive = false;
+			fireActive = true;
+			timeRemainingTilesActive = timeTilesActive;
+			firstTimeFireActive = true;
+		}
+	}
+
+	if (fireActive) {
+		if (firstTimeFireActive) {
+			UpdateFireActiveTiles(true);
+			firstTimeFireActive = false;
+		}
+
+		if (timeRemainingTilesActive > 0.f) {
+			timeRemainingTilesActive -= Time::GetDeltaTime();
+		}
+		else {
+			UpdateFireActiveTiles(false);
+			if (timerTilesClosingRemaining <= timerTilesClosing) {
+				timerTilesClosingRemaining += Time::GetDeltaTime();
+			}
+			else {
+				warningActive = true;
+				fireActive = false;
+				patternFinished = true;
+				previousPattern = currentPattern;
+				timeRemainingWarning = timeWarning;
+				firstTimeWarning = true;
+				timerTilesClosingRemaining = 0.f;
+			}
+		}
+
+	}
+
+}
+
+void FloorIsLava::UpdateWarningTiles(bool activate)
+{
+	//corridor
+	for (int i = 0; i < CORRIDOR_TILES; ++i) {
```

Comment:
This could be a generic method in which you pass the parameters of the corridor and the arena